Support `compute build` for 'other' language option using `[scripts.build]`

Integralist opened the pull request:
Currently a project that uses an unsupported C@E language can't use the `compute build` command (which means having to do things like using `compute serve --skip-build`). The new `[scripts.build]` configuration in the fastly.toml manifest helps us to better support `compute build` for unofficial languages.

## EXAMPLE

```toml
[scripts]
build = "mkdir -p bin && tinygo build -target wasi -o bin/main.wasm main.go && fastly compute pack --wasm-binary bin/main.wasm"
```

I was able to run `fastly compute serve` and have it 'build' and 'serve' locally via Viceroy.

kailan reviewed (approved):
Love this!

Integralist commented:
If this goes ahead, then we should update [our documentation](https://developer.fastly.com/learning/compute/custom/) and also the `compute init` output so they reflect this change.

As far as [our documentation](https://developer.fastly.com/learning/compute/custom/) is concerned I think it can probably stay the same but have something like the following appended:

> **HINT:** The use of `compute pack` can be replaced by adding a `[scripts.build]` setting within your [`fastly.toml` manifest](https://developer.fastly.com/reference/compute/fastly-toml/).

Currently the `compute init` output shows...

```
Initialized package testing-cli to:
        /Users/integralist/Code/go/testing-cli

To package a pre-compiled Wasm binary for deployment, run:
        fastly compute pack

To deploy the package, run:
        fastly compute deploy
```

I imagine this should change back to the normal output:

```
Initialized package testing-cli to:
        /Users/integralist/Code/go/testing-cli

To publish the package (build and deploy), run:
        fastly compute publish
```

Because if the user of an 'other' language project tries to build without doing anything, then they'll now see (as part of this PR) the error...

```
$ fastly compute build
✓ Initializing...
✓ Verifying package manifest...
✓ Verifying local other toolchain...
✗ Building package using other toolchain...

ERROR: error reading custom build instructions from fastly.toml manifest.

Add a [scripts.build] setting for your custom build process. See more at https://developer.fastly.com/reference/fastly-toml/
```

/cc @triblondon

fgsch commented:
Not specific to this PR but I wonder if we want to let the user know what command we are going to run and give the user the option to abort it or require some flag to continue.  

My thinking here is that a third party could share a project with a build command for language that would normally build without it (e.g. Rust), and do evil things.

Integralist commented:
@fgsch I think your suggestion is valid and I'll look to add something to that effect.

Integralist commented:
@fgsch @kailan 

As requested I've added a prompt which informs the user a custom build has been configured (in case they were not aware of it, or what it was doing), and I ask them to confirm they want to go ahead with the build.

I've also added a new `--accept-custom-build` flag to `compute build` so that this prompt can be ignored.

Integralist commented:
@fgsch @kailan may I ask you both re-review, just to be sure you're still happy with the changes.

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,24 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+	// language that wouldn't normally require one (e.g. Rust), and do evil
+	// things. So we should notify the user and confirm they would like to
+	// continue with the build.
+	if !c.Flags.AcceptCustomBuild && language.Name == "other" && toolchain == "custom" {
```
Wouldn't `language.Name == "other"` limit the check to other languages?

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,24 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+			text.Info(out, "Stopping the build process.")
+			return nil
+		}
 	}
```
Would it be too much to always display `This project has a custom build script defined in the fastly.toml manifest` in verbose mode?

Integralist commented on pkg/commands/compute/build_test.go:
```diff
@@ -532,3 +533,139 @@ func TestBuildJavaScript(t *testing.T) {
+		wantOutput           []string
+		wantRemediationError string
+	}{
+		{
```
done

fgsch commented on pkg/commands/compute/build_test.go:
```diff
@@ -532,3 +533,152 @@ func TestBuildJavaScript(t *testing.T) {
+			},
+			wantError:            "error reading custom build instructions from fastly.toml manifest",
+			wantRemediationError: "Add a [scripts.build] setting for your custom build process",
+		},
```
Could we also add a test for a non-other language?

Integralist replied:
done

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,29 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+		if c.Globals.Flag.Verbose {
+			text.Info(out, customBuildMsg)
+			customBuildMsg = ""
+		}
```
I wonder if we should always display this in verbose mode for custom build scripts, regardless of the AcceptCustomBuild flag.

Integralist replied:
Have fixed locally, will push up shortly

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,29 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+			text.Info(out, customBuildMsg)
+			customBuildMsg = ""
+		}
+		label := fmt.Sprintf("%sAre you sure you want to continue with the build step? [y/N] ", customBuildMsg)
```
Not a blocker: for symmetry, can we check if we printed customBuildMsg above and avoid printing it again?

fgsch replied:
Nevermind this, it's already covered.

fgsch commented on pkg/commands/compute/language_other.go:
```diff
@@ -0,0 +1,69 @@
+	s := fstexec.Streaming{
+		Command: cmd,
+		Args:    args,
+		Env:     []string{},
```
Question (not a blocker): Any reason we are not passing the existing environment?

Integralist replied:
I'll update this in a separate PR

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,31 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
 
-	if c.Manifest.File.Scripts.Build != "" {
-		lang = "custom"
+	customBuildMsg := "This project has a custom build script defined in the fastly.toml manifest. "
```
nitpick (not a blocker): add the newline to customBuildMsg (and zap the trailing space), remove the newline from the fmt.Sprintf() below so the prompt further below becomes:

```
This project has a custom build script defined in the fastly.toml manifest.
Are you sure you want to continue with the build step? [y/N] 
```

fgsch reviewed (approved):
I've left some non-blocking comments but this looks good to me and any remaining things can be addressed in future PRs if needed.

I still need to test it but I trust your judgement and I will let you make the call.

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,31 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
-		lang = "custom"
+	customBuildMsg := "This project has a custom build script defined in the fastly.toml manifest. "
+
+	if toolchain == "custom" && c.Globals.Flag.Verbose {
+		text.Info(out, fmt.Sprintf("%s\n\n%s\n\n", customBuildMsg, c.Manifest.File.Scripts.Build))
+		customBuildMsg = ""
+	}
```
```suggestion
	if toolchain == "custom" && (c.Globals.Flag.Verbose || !c.Flags.AcceptCustomBuild) {
		text.Info(out, fmt.Sprintf("%s\n\n%s\n\n", customBuildMsg, c.Manifest.File.Scripts.Build))
	}
```

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -151,11 +169,31 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+	// things. So we should notify the user and confirm they would like to
+	// continue with the build.
+	if toolchain == "custom" && !c.Flags.AcceptCustomBuild {
+		label := fmt.Sprintf("%sAre you sure you want to continue with the build step? [y/N] ", customBuildMsg)
```
```suggestion
		label := fmt.Sprintf("Are you sure you want to continue with the build step? [y/N] ")
```

fgsch reviewed (approved):
Thank you. This is great! 🙏

fgsch commented on pkg/commands/compute/build.go:
```diff
@@ -192,7 +192,8 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
 				text.Info(out, "Stopping the build process.")
-				return nil
+				text.Break(out)
+				return fsterr.ErrBuildStopped
```
Is there any way to test this case?

Integralist replied:
I actually had to update the tests to make them pass by ensuring I set an expectation for an error to be returned (see my last commit)

fgsch reviewed (approved):
👍
